[1/N] Decouple Flight Recorder from NCCL utils

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141648

Part of the effort to make Flight Recorder device agnostic.

Step 1: Move it out of NCCLUtils.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
[1/N] Decouple Flight Recorder from NCCL utils

[ghstack-poisoned]

kwen2501 committed:
Update on "[1/N] Decouple Flight Recorder from NCCL utils"


Part of the effort to make Flight Recorder device agnostic.

Step 1: Move it out of NCCLUtils.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/FlightRecorder.hpp:
```diff
@@ -0,0 +1,257 @@
+// (minor when adding fields, major when changing existing fields)
+// Also update both JSON and Pickle dumps to make use of the newly defined
+// field(s).
+DEFINE_CONSTANT(version_val, "2.4")
```
Can you bump up the version to maybe "3.0", so that we know the dump before and after this change since this is a big change.

kwen2501 replied:
Hmm, there is no substantial value change here.

kwen2501 committed:
Update on "[1/N] Decouple Flight Recorder from NCCL utils"


Part of the effort to make Flight Recorder device agnostic.

Step 1: Move it out of NCCLUtils.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge -i

kwen2501 commented:
@pytorchbot merge -f "Unrelated doc build failure: Undocumented Python objects: torch.fx.experimental.symbolic_shapes, Functions: * is_concrete_float"

kwen2501 commented:
Internal forward fix: [D66563186](https://www.internalfb.com/diff/D66563186)
